File: src/edge.cpp

#include "edge.hpp"
#include "spdlog/spdlog.h"

#include <sstream>


Edge::Edge(Point* left, Point* right){
    m_left_target = left;
    m_right_target = right;
	m_he_right = new HalfEdge(right);
	m_he_left = new HalfEdge(left);
	m_he_right->opposite() = m_he_left;
	m_he_left->opposite() = m_he_right;
	m_start = m_end = 0;
    spdlog::debug("edge - created {0}", asString());
}
std::string Edge::asString(){ 
	std::stringstream ss;
	ss << (m_start ? m_start->asString() : "(0)") << "--" << (m_end ? m_end->asString() : "(0)")
	    << ", tl=" << m_left_target->asString() << ", tr=" << m_right_target->asString();
	return "E(" + ss.str() + ")"; // + m_left_target->asString() + " | " + m_right_target->asString() + ")"; 
};
void Edge::setStart(Point* p){ 
    spdlog::debug("edge - start {0}--(0)", p->asString());
	m_start = p; };
void Edge::setEnd(Point* p){ 
	spdlog::debug("edge - end {0}--{1}", m_start ? m_start->asString() : "(0)", p->asString());
	m_end = p; };

HalfEdge* const & Edge::he_right() const { return m_he_right; };
HalfEdge*       & Edge::he_right() { return m_he_right; };
HalfEdge* const & Edge::he_left() const { return m_he_left; };
HalfEdge*       & Edge::he_left() { return m_he_left; };

Point* const & Edge::leftTarget() const { return m_left_target; };
Point*       & Edge::leftTarget(){ return m_left_target; };
Point* const & Edge::rightTarget() const{ return m_right_target; };
Point*       & Edge::rightTarget(){ return m_right_target; };
Point* const & Edge::startPoint() const{ return m_start; };
Point*       & Edge::startPoint(){ return m_start; };
Point* const & Edge::endPoint() const{ return m_end; };
Point*       & Edge::endPoint(){ return m_end; };

File: src/Edge.py
import logging
from src.Point import Point
from lib.voronoi_lib import Edge as Edge_cpp

logger = logging.getLogger()


class Edge:
    """data object wrapper for a cpp Edge"""

    def __init__(self, e: Edge_cpp):
        self.e = e
        self.left_target = Point(e.left_target().x(), e.left_target().y()) if e.left_target() else None
        self.right_target = Point(e.right_target().x(), e.right_target().y()) if e.right_target() else None
        self.start_point = Point(e.start_point().x(), e.start_point().y()) if e.start_point() else None
        self.end_point = Point(e.end_point().x(), e.end_point().y()) if e.end_point() else None
        logger.debug(f"created {str(self)}")

    def nonzero_length(self):
        if self.start_point and self.end_point:
            return self.start_point != self.end_point
        return True

    def for_plot(self):
        x1, y1 = self.start_point.coordinates()
        x2, y2 = self.end_point.coordinates()
        return x1, y1, x2, y2

    def get_normal(self):
        """computes normal of vector between left and right target"""
        x1, y1 = self.right_target.coordinates()
        x2, y2 = self.left_target.coordinates()
        mx = (x1 + x2) / 2
        my = (y1 + y2) / 2
        if self.start_point:
            x, y = self.start_point.coordinates()
        elif self.end_point:
            x, y = self.end_point.coordinates()
        else:
            logger.warning("no start or endpoint")
        xn, yn = mx - x, my - y
        return xn, yn

    def get_unbound_start(self):
        """returns starting point of the unbound line"""
        if self.start_point and self.end_point:
            return None
        if self.start_point:
            return self.start_point
        else:
            return self.end_point

    def bound(self, x, y):
        """bound the edge by the given point"""
        if self.start_point and self.end_point:
            logger.warning(f"edge was already bound, tried ({x}, {y}) for {str(self)}")
        if self.start_point:
            self.end_point = Point(x, y)
        else:
            self.start_point = Point(x, y)
        logger.debug(f"bound {str(self)}")

    def __str__(self):
        return f"Edge ({str(self.start_point)} <-> {str(self.end_point)}), tl={str(self.left_target)}, tr={str(self.right_target)}"